src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp

```diff
@@ -0,0 +1,172 @@
+      0.5 * (square(get<0>(x_minus_center)) + square(get<1>(x_minus_center)) +
+             square(get<2>(x_minus_center)) - a_squared);
+  const DataVector r_squared =
+      half_xsq_minus_asq + sqrt(square(half_xsq_minus_asq) + a_dot_x_squared);
+  const DataVector a_dot_x_over_rsquared = a_dot_x / r_squared;
+
+  const DataVector deriv_log_r_denom = 0.5 / (r_squared - half_xsq_minus_asq);
+  tnsr::i<DataVector, 3> deriv_log_r;
+  for (size_t i = 0; i < 3; ++i) {
+    deriv_log_r.get(i) =
+        deriv_log_r_denom *
+        (x_minus_center.get(i) + gsl::at(spin_, i) * a_dot_x_over_rsquared);
+  }
+
+  const DataVector H_denom = 1.0 / (square(r_squared) + a_dot_x_squared);
+  const DataVector H = mass_ * sqrt(r_squared) * r_squared * H_denom;
+
+  tnsr::i<DataVector, 3> dH;
+  {
+    const DataVector fac1 = H * (3.0 - 4.0 * square(r_squared) * H_denom);
+    const DataVector fac2 = H * (2.0 * H_denom * a_dot_x);
+    for (size_t i = 0; i < 3; ++i) {
+      dH.get(i) = fac1 * deriv_log_r.get(i) - fac2 * gsl::at(spin_, i);
+    }
+  }
+
+  tnsr::i<DataVector, 3> a_cross_x;
+  get<0>(a_cross_x) =
+      spin_[1] * get<2>(x_minus_center) - spin_[2] * get<1>(x_minus_center);
+  get<1>(a_cross_x) =
+      spin_[2] * get<0>(x_minus_center) - spin_[0] * get<2>(x_minus_center);
+  get<2>(a_cross_x) =
+      spin_[0] * get<1>(x_minus_center) - spin_[1] * get<0>(x_minus_center);
+
+  const DataVector denom = 1.0 / (r_squared + a_squared);
+  const DataVector r = sqrt(r_squared);
+
+  tnsr::i<DataVector, 3> null_form;
+  {
+    const DataVector fac = a_dot_x / r;
+    for (size_t i = 0; i < 3; ++i) {
+      null_form.get(i) = denom * (r * x_minus_center.get(i) - a_cross_x.get(i) +
+                                  fac * gsl::at(spin_, i));
+    }
+  }
+
+  tnsr::ij<DataVector, 3> deriv_null_form;
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j < 3; j++) {
+      deriv_null_form.get(j, i) =
+          denom * (gsl::at(spin_, i) * gsl::at(spin_, j) / r +
+                   (x_minus_center.get(i) - 2.0 * r * null_form.get(i) -
+                    a_dot_x_over_rsquared * gsl::at(spin_, i)) *
+                       deriv_log_r.get(j) * r);
+      if (i == j) {
+        deriv_null_form.get(j, i) += denom * r;
+      } else {  //  add denom*epsilon^ijk a_k
+        size_t k = (j + 1) % 3;
+        if (k == i) {  // j+1 = i (cyclic), so choose minus sign
+          k++;
+          k = k % 3;  // and set k to be neither i nor j
+          deriv_null_form.get(j, i) -= denom * gsl::at(spin_, k);
+        } else {  // i+1 = j (cyclic), so choose plus sign
+          deriv_null_form.get(j, i) += denom * gsl::at(spin_, k);
+        }
+      }
+    }
+  }
+
+  const constexpr double null_vector_0 = -1.0;
+  const DataVector lapse_squared = 1.0 / (1. + 2. * H * square(null_vector_0));
+
+  auto result = make_with_value<Variables<KerrSchild::Tags>>(x, 0.0);
+
+  get(get<lapse>(result)) = sqrt(lapse_squared);
+
+  {
+    const DataVector fac =
```

Comment:
Since this is only used in one calculation should it just be inlined to prevent the memory allocation?

Reply:
Don't know.  I kept it out because if inlined it would be computed 3 times instead of once.

Reply:
Oh yeah of course, nevermind.